bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheAsync.java

```diff
@@ -0,0 +1,137 @@
+     */
+    public static interface ExpiringCacheUpdate {
+        void requestCacheUpdate();
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param expiry the duration in milliseconds for how long the value stays valid. Must be greater than 0.
+     * @param cacheUpdater The cache will use this callback if a new value is needed. Must not be null.
+     * @throws IllegalArgumentException For an expire value <=0 or a null cacheUpdater.
+     */
+    public ExpiringCacheAsync(long expiry, ExpiringCacheUpdate cacheUpdater) throws IllegalArgumentException {
+        this.expiry = TimeUnit.MILLISECONDS.toNanos(expiry);
+        this.cacheUpdater = cacheUpdater;
+        if (expiry <= 0) {
+            throw new IllegalArgumentException("Cache expire time must be greater than 0");
+        }
+        if (cacheUpdater == null) {
+            throw new IllegalArgumentException("A cache updater is necessary");
+        }
+    }
+
+    /**
+     * Returns the value - possibly from the cache, if it is still valid.
+     *
+     * @return the value
+     */
+    public void getValue(Consumer<V> callback) {
+        if (isExpired()) {
+            refreshValue(callback);
+        } else {
+            callback.accept(value);
+        }
+    }
+
+    /**
+     * Invalidates the value in the cache.
+     */
+    public void invalidateValue() {
+        expiresAt = 0;
+    }
+
+    /**
+     * Updates the cached value with the given one.
+     *
+     * @param newValue The new value. All listeners, registered by getValueAsync() and refreshValue(), will be notified
+     *            of the new value.
+     */
+    public void setValue(V newValue) {
+        refreshRequested = false;
+        value = newValue;
+        expiresAt = getCurrentNanoTime() + expiry;
+        // Inform all callback handlers of the new value and clear the list
+        for (Consumer<V> callback : waitingCacheCallbacks) {
+            callback.accept(value);
+        }
+        waitingCacheCallbacks.clear();
+    }
+
+    /**
+     * Returns an arbitrary time reference in nanoseconds.
+     * This is used for the cache to determine if a value has expired.
+     */
+    public long getCurrentNanoTime() {
+        return System.nanoTime();
+    }
+
+    /**
+     * Refreshes and returns the value asynchronously.
+     *
+     * @return the new value
+     */
+    public void refreshValue(Consumer<V> callback) {
+        waitingCacheCallbacks.add(callback);
+        if (refreshRequested) {
+            return;
+        }
+        refreshRequested = true;
+        expiresAt = 0;
+        cacheUpdater.requestCacheUpdate();
+    }
+
+    /**
+     * Checks if the value is expired.
+     *
+     * @return true if the value is expired
+     */
+    public boolean isExpired() {
+        return expiresAt < getCurrentNanoTime();
+    }
+
+    /**
+     * Return the raw value, no matter if it is already
+     * expired or still valid.
+     */
+    public V getExpiredValue() {
```

Comment:
Yup, I can do that. It's a matter of taste I guess. I wanted to make it very obvious that the value might be expired.